### DISCUSSION TITLE
Pyright reports false positive error with async streams

### DISCUSSION THREAD

**Discussion by RobertCraigie** [encode/httpx#1829] [Potential Issue] [answered]
## Problem

```py
from httpx import AsyncClient

async def main(url: str) -> None:
    client = AsyncClient()

    async with client.stream('GET', url) as resp:
        ...
```

Pyright reports:

```
error: Object of type "AsyncIterator[Response]" cannot be used with "with" because it does not implement __aenter__ (reportGeneralTypeIssues)
```

## Solution

Although I haven't checked, I am fairly certain that this issue can be solved by changing how `httpx/_compat.py` imports the async context manager:

```py
import sys

# `contextlib.asynccontextmanager` exists from Python 3.7 onwards.
# For 3.6 we require the `async_generator` package for a backported version.
if sys.version_info >= (3, 7):
    from contextlib import asynccontextmanager
else:
    from async_generator import asynccontextmanager
```

**Comment by lovelydinosaur** [comment, marked as answer]
>  I am fairly certain that this issue can be solved by changing how httpx/_compat.py imports the async context manager

Okay, I'm curious... why *might* that satisfy `pyright`'s type checking, where the `try... except ImportError ...` style would not?

```python
# `contextlib.asynccontextmanager` exists from Python 3.7 onwards.
# For 3.6 we require the `async_generator` package for a backported version.
try:
    from contextlib import asynccontextmanager  # type: ignore
except ImportError:
    from async_generator import asynccontextmanager  # type: ignore # noqa
```

If you're able to verify that assumption then I guess we *could* consider if it's worth switch our style for conditional imports.

**Reply by HarrySky** [reply]
> Okay, I'm curious... why *might* that satisfy `pyright`'s type checking, where the `try... except ImportError ...` style would not?

Because pyright is static type checking tool that does not run Python code in order to catch that `ImportError`

Had the same problem with pyright, so now I just use `if sys.version_info ...` check. Can't find the issue I stumbled upon back then to link here though.

**Reply by lovelydinosaur** [reply]
> Because pyright is static type checking tool that does not run Python code in order to catch that ImportError

Okay. Does it recognise `sys.version_info` branching, then?

**Reply by HarrySky** [reply]
Yes, if you write it as `if sys.version_info < (3, 8)`.

AFAIK it will not work if you use `from sys import version_info`

**Reply by lovelydinosaur** [reply]
Okeydokes - I'd be happy to accept a pull request tweaking the conditional import there, then.

**Reply by HarrySky** [reply]
@RobertCraigie want to send a PR fixing your issue?

**Reply by RobertCraigie** [reply]
> @RobertCraigie want to send a PR fixing your issue?

Yes, I'll send a PR as soon as possible.
